chapter10/mssw/exercise4.pl

```diff
@@ -0,0 +1,59 @@
+#!/usr/bin/perl
+
+use strict;
+use warnings;
+
+use Data::Dumper;
+
+sub data_for_path {
+    my $path = shift;
+    if ( -f $path || -l $path ) {
+        return undef;
+    }
+
+    if ( -d $path ) {
+        my %directory;
+        opendir PATH, $path or die "Cannot opendir $path: $!";
+        my @names = readdir PATH;
+        closedir PATH;
+        foreach my $name (@names) {
+            next if $name eq '.' or $name eq '..';
+            $directory{$name} = data_for_path("$path/$name");
+        }
+        return \%directory;
+    }
+    warn "$path is neither a file nor a directory\n";
+    return undef;
+}
+
+sub dump_data_for_path {
+    my ( $path, $data, $indent ) = @_;
+    $indent //= 0;
```

Comment:
Checking this every time seems a bit unnecessary, could maybe just seed the first call with an indent.